invokeai/app/services/model_records/model_records_sql.py

```diff
@@ -0,0 +1,396 @@
+                """--sql
+                SELECT config FROM model_config
+                WHERE id=?;
+                """,
+                (key,),
+            )
+            rows = self._cursor.fetchone()
+            if not rows:
+                raise UnknownModelException("model not found")
+            model = ModelConfigFactory.make_config(json.loads(rows[0]))
+        return model
+
+    def exists(self, key: str) -> bool:
+        """
+        Return True if a model with the indicated key exists in the databse.
+
+        :param key: Unique key for the model to be deleted
+        """
+        count = 0
+        with self._lock:
+            try:
+                self._cursor.execute(
+                    """--sql
+                    select count(*) FROM model_config
+                    WHERE id=?;
+                    """,
+                    (key,),
+                )
+                count = self._cursor.fetchone()[0]
+            except sqlite3.Error as e:
+                raise e
+        return count > 0
+
+    def search_by_name(
+        self,
+        model_name: Optional[str] = None,
+        base_model: Optional[BaseModelType] = None,
+        model_type: Optional[ModelType] = None,
+    ) -> List[AnyModelConfig]:
+        """
+        Return models matching name, base and/or type.
+
+        :param model_name: Filter by name of model (optional)
+        :param base_model: Filter by base model (optional)
+        :param model_type: Filter by type of model (optional)
+
+        If none of the optional filters are passed, will return all
+        models in the database.
+        """
+        results = []
+        where_clause = []
+        bindings = []
+        if model_name:
+            where_clause.append("model_name=?")
+            bindings.append(model_name)
+        if base_model:
+            where_clause.append("base_model=?")
+            bindings.append(base_model)
+        if model_type:
+            where_clause.append("model_type=?")
+            bindings.append(model_type)
+        where = f"WHERE {' AND '.join(where_clause)}" if where_clause else ""
+        with self._lock:
+            try:
+                self._cursor.execute(
+                    f"""--sql
+                    select config FROM model_config
+                    {where};
+                    """,
+                    tuple(bindings),
+                )
+                results = [ModelConfigFactory.make_config(json.loads(x[0])) for x in self._cursor.fetchall()]
+            except sqlite3.Error as e:
+                raise e
+        return results
+
+    def search_by_path(self, path: Union[str, Path]) -> List[ModelConfigBase]:
+        """Return models with the indicated path."""
+        results = []
+        with self._lock:
+            try:
+                self._cursor.execute(
+                    """--sql
+                    SELECT config FROM model_config
+                    WHERE model_path=?;
```

Comment:
You will probably want to add an index on the table for any fields that you can search on. May not matter as much for OSS unless someone has a ton of models

Reply:
I've frequently been advised to wait to see where the performance bottlenecks are before adding indexes, but happy to do this.